File: jvm-experiments-timeperiod/src/main/java/kr/kth/timeperiod/TimeInterval.java
package kr.kth.timeperiod;

import kr.kth.commons.Guard;
import kr.kth.timeperiod.timerange.TimeRange;
import kr.kth.timeperiod.tools.TimeTool;
import lombok.extern.slf4j.Slf4j;
import org.joda.time.DateTime;

import static kr.kth.commons.Guard.shouldNotBeNull;

/**
 * 설명을 추가하세요.
 * User: sunghyouk.bae@gmail.com
 * Date: 12. 12. 20
 */
@Slf4j
public class TimeInterval extends TimePeriodBase implements ITimeInterval {

    private static final long serialVersionUID = -5917038576117115113L;

    public static TimeInterval Anytime = new TimeInterval(TimeSpec.MinPeriodTime,
            TimeSpec.MaxPeriodTime,
            IntervalEdge.Closed,
            IntervalEdge.Closed,
            false,
            true);

    public static TimeRange toTimeRange(ITimeInterval interval) {
        if (interval == null)
            return null;

        return interval.isAnyTime()
                ? TimeRange.Anytime
                : new TimeRange(interval.getStartInterval(),
                interval.getEndInterval(),
                interval.isReadonly());
    }

    public static TimeBlock toTimeBlock(ITimeInterval interval) {
        if (interval == null)
            return null;
        return interval.isAnyTime()
                ? TimeBlock.Anytime
                : new TimeBlock(interval.getStartInterval(),
                interval.getEndInterval(),
                interval.isReadonly());
    }

    // region << constructors >>

    public TimeInterval() {
        this(TimeSpec.MinPeriodTime, TimeSpec.MaxPeriodTime);
    }

    public TimeInterval(DateTime moment) {
        this(moment, IntervalEdge.Closed, IntervalEdge.Closed, true, false);
    }

    public TimeInterval(DateTime moment, IntervalEdge startEdge, IntervalEdge endEdge) {
        this(moment, startEdge, endEdge, true, false);
    }

    public TimeInterval(DateTime moment, IntervalEdge startEdge, IntervalEdge endEdge, boolean intervalEnabled, boolean readonly) {
        this(moment, moment, startEdge, endEdge, intervalEnabled, readonly);
    }

    public TimeInterval(DateTime startInterval, DateTime endInterval) {
        this(startInterval, endInterval, IntervalEdge.Closed, IntervalEdge.Closed, true, false);
    }

    public TimeInterval(DateTime startInterval, DateTime endInterval, IntervalEdge startEdge, IntervalEdge endEdge) {
        this(startInterval, endInterval, startEdge, endEdge, true, false);
    }

    public TimeInterval(DateTime startInterval, DateTime endInterval, IntervalEdge startEdge, IntervalEdge endEdge,
                        boolean intervalEnabled, boolean readonly) {
        super(startInterval, endInterval, readonly);

        this.startEdge = startEdge;
        this.endEdge = endEdge;
        this.intervalEnabled = intervalEnabled;
    }

    public TimeInterval(ITimePeriod src) {
        super(src);

        if (src instanceof ITimeInterval) {
            ITimeInterval interval = (ITimeInterval) src;

            this.start = interval.getStartInterval();
            this.end = interval.getEndInterval();
            this.startEdge = interval.getStartEdge();
            this.endEdge = interval.getEndEdge();
            this.intervalEnabled = interval.isIntervalEnabled();
        }
    }

    // endregion

    private boolean intervalEnabled;
    private IntervalEdge startEdge;
    private IntervalEdge endEdge;

    @Override
    public boolean isStartOpen() {
        return (this.startEdge == IntervalEdge.Open);
    }

    @Override
    public boolean isEndOpen() {
        return (this.endEdge == IntervalEdge.Open);
    }

    @Override
    public boolean isStartClosed() {
        return (this.startEdge == IntervalEdge.Closed);
    }

    @Override
    public boolean isEndClosed() {
        return (this.endEdge == IntervalEdge.Closed);
    }

    @Override
    public boolean isClosed() {
        return isStartClosed() && isEndClosed();
    }

    @Override
    public boolean isEmpty() {
        return isMoment() && !isClosed();
    }

    @Override
    public boolean isDegenerate() {
        return isMoment() && isClosed();
    }

    @Override
    public boolean isIntervalEnabled() {
        return intervalEnabled;
    }

    @Override
    public void setIntervalEnabled(boolean intervalEnabled) {
        assertMutable();
        this.intervalEnabled = intervalEnabled;
    }

    @Override
    public boolean hasStart() {
        return (this.start != TimeSpec.MinPeriodTime) || !isStartClosed();
    }

    @Override
    public DateTime getStartInterval() {
        return this.start;
    }

    @Override
    public void setStartInterval(DateTime start) {
        assertMutable();
        Guard.shouldBe(start.compareTo(this.end) <= 0,
                "새로운 getStart=[%s]는 getEnd=[%s] 보다 작거나 같아야 합니다.", start, this.end);
        this.start = start;
    }

    @Override
    public DateTime getStart() {
        if (this.intervalEnabled && isStartOpen())
            return this.start.plus(1L);
        return this.start;
    }

    @Override
    public IntervalEdge getStartEdge() {
        return this.startEdge;
    }

    @Override
    public void setStartEdge(IntervalEdge edge) {
        assertMutable();
        this.startEdge = edge;
    }

    @Override
    public boolean hasEnd() {
        return this.end != TimeSpec.MaxPeriodTime || !isEndClosed();
    }


    @Override
    public DateTime getEndInterval() {
        return this.end;
    }

    @Override
    public void setEndInterval(DateTime end) {
        assertMutable();
        Guard.shouldBe(end.compareTo(this.start) >= 0,
                "새로운 getEnd=[%s]가 getStart=[{}]보다 크거나 같아야 합니다.", end, this.start);
        this.end = end;
    }

    @Override
    public DateTime getEnd() {
        if (intervalEnabled || isEndOpen())
            return this.end.minus(1L);
        return this.end;
    }

    @Override
    public IntervalEdge getEndEdge() {
        return endEdge;
    }

    @Override
    public void setEndEdge(IntervalEdge edge) {
        assertMutable();
        this.endEdge = edge;
    }

    @Override
    public Long getDuration() {
        return this.end.minus(this.start.getMillis()).getMillis();
    }

    @Override
    public void setup(DateTime start, DateTime end) {
        super.setup(start, end);

        if (TimeInterval.log.isDebugEnabled())
            TimeInterval.log.debug("기간을 다시 설정합니다. getStart=[{}], getEnd=[{}]", start, end);

        assertMutable();

        TimeTool.adjustPeriod(start, end);

        this.start = Guard.firstNotNull(start, TimeSpec.MinPeriodTime);
        this.end = Guard.firstNotNull(end, TimeSpec.MaxPeriodTime);
    }

    @Override
    public void expandStartTo(DateTime moment) {
        assertMutable();
        if (start.compareTo(moment) > 0)
            start = moment;
    }

    @Override
    public void expandEndTo(DateTime moment) {
        assertMutable();
        if (end.compareTo(moment) < 0)
            end = moment;
    }

    @Override
    public void expandTo(DateTime moment) {
        shouldNotBeNull(moment, "moment");
        assertMutable();

        expandStartTo(moment);
        expandEndTo(moment);
    }

    @Override
    public void expandTo(ITimePeriod period) {
        shouldNotBeNull(period, "period");
        assertMutable();

        expandStartTo(period.getStart());
        expandEndTo(period.getEnd());
    }

    @Override
    public void shrinkStartTo(DateTime moment) {
        shouldNotBeNull(moment, "moment");
        assertMutable();

        if (start.compareTo(moment) < 0)
            start = moment;
    }

    @Override
    public void shrinkEndTo(DateTime moment) {
        shouldNotBeNull(moment, "moment");
        assertMutable();

        if (end.compareTo(moment) > 0)
            end = moment;
    }

    @Override
    public void shrinkTo(DateTime moment) {
        shouldNotBeNull(moment, "moment");
        assertMutable();

        shrinkStartTo(moment);
        shrinkEndTo(moment);
    }

    @Override
    public void shrinkTo(ITimePeriod period) {
        shouldNotBeNull(period, "period");
        assertMutable();

        shrinkStartTo(period.getStart());
        shrinkEndTo(period.getEnd());
    }
}

File: jvm-experiments-timeperiod/src/main/scala/kr/kth/timeperiod/TimeInterval.scala
package kr.kth.timeperiod

import kr.kth.commons.Guard
import kr.kth.commons.slf4j.Logging
import kr.kth.commons.tools.ScalaHash
import kr.kth.timeperiod.IntervalEdge._
import org.joda.time.DateTime

/**
 * kr.kth.timeperiod.TimeInterval
 * User: sunghyouk.bae@gmail.com
 * Date: 13. 1. 14
 */
class TimeInterval(start: DateTime,
                   end: DateTime,
                   startEdge: IntervalEdge = IntervalEdge.Closed,
                   endEdge: IntervalEdge = IntervalEdge.Closed,
                   intervalEnabled: Boolean = true,
                   readonly: Boolean = false)
  extends TimePeriod(start, end, readonly) with ITimeInterval {

  _startEdge = startEdge
  _endEdge = endEdge
  _intervalEnabled = intervalEnabled

}

object TimeInterval extends Logging {

  lazy val Anytime = apply(readonly = true)

  def apply(readonly: Boolean) =
    apply(start = TimeSpec.MinPeriodTime,
          end = TimeSpec.MaxPeriodTime,
          readonly = readonly)

  def apply(moment: DateTime,
            startEdge: IntervalEdge = IntervalEdge.Closed,
            endEdge: IntervalEdge = IntervalEdge.Closed,
            intervalEnabled: Boolean = true,
            readonly: Boolean = true) =
    apply(moment,
          moment,
          startEdge,
          endEdge,
          intervalEnabled,
          readonly)

  def apply(start: DateTime,
            end: DateTime,
            startEdge: IntervalEdge = IntervalEdge.Closed,
            endEdge: IntervalEdge = IntervalEdge.Closed,
            intervalEnabled: Boolean = true,
            readonly: Boolean = true) =
    new TimeInterval(start, end, startEdge, endEdge, intervalEnabled, readonly)

  def apply(start: DateTime,
            duration: Long,
            startEdge: IntervalEdge = IntervalEdge.Closed,
            endEdge: IntervalEdge = IntervalEdge.Closed,
            intervalEnabled: Boolean = true,
            readonly: Boolean = true) =
    apply(start, start.plus(duration), startEdge, endEdge, intervalEnabled, readonly)

  def apply(source: ITimePeriod) = {
    Guard.shouldNotBeNull(source, "source")
    if (source == Anytime) Anytime
    else if (source.isInstanceOf[ITimeInterval]) {
      val interval = source.asInstanceOf[ITimeInterval]
      apply(interval.getStart,
            interval.getEnd,
            interval.getStartEdge,
            interval.getEndEdge,
            interval.isIntervalEnabled,
            interval.isReadonly)
    } else {
      apply(source.getStart, source.getEnd, readonly = source.isReadonly)
    }
  }

  def apply(source: ITimePeriod, readonly: Boolean) = {
    Guard.shouldNotBeNull(source, "source")

    if (source.isInstanceOf[ITimeInterval]) {
      val interval = source.asInstanceOf[ITimeInterval]
      apply(interval.getStart,
            interval.getEnd,
            interval.getStartEdge,
            interval.getEndEdge,
            interval.isIntervalEnabled,
            readonly)
    } else {
      apply(source.getStart, source.getEnd, readonly = readonly)
    }
  }

  def unapply(range: ITimeInterval) =
    (range.getStart, range.getEnd, range.getStartEdge, range.getEndEdge, range.isIntervalEnabled, range.isReadonly)
}

trait ITimeInterval extends ITimePeriod {

  protected var _intervalEnabled = false
  protected var _startEdge = IntervalEdge.Closed
  protected var _endEdge = IntervalEdge.Closed

  /**
   * 시작 시각이 개구간인지 여부
   */
  def isStartOpen: Boolean = _startEdge == IntervalEdge.Open

  /**
   * 완료 시각이 개구간인지 여부
   */
  def isEndOpen: Boolean = _endEdge == IntervalEdge.Open

  /**
   * 시작시각이 폐구간인지 여부
   *
   * @return 시작시각이 폐구간이면 true, 개구간이면 false를 반환
   */
  def isStartClosed: Boolean = _start == IntervalEdge.Closed

  /**
   * 완료시각이 폐구간인지 여부
   *
   * @return 완료시각이 폐구간이면 true, 개구간이면 false를 반환
   */
  def isEndClosed: Boolean = _end == IntervalEdge.Closed

  /**
   * 기간이 폐구간인지 여부
   */
  def isClosed: Boolean = isStartClosed && isEndClosed

  /**
   * 기간이 비었는가 여부 (시작시각과 완료시각이 같으면 기간은 비었다)
   */
  def isEmpty: Boolean = isMoment && !isClosed

  /**
   * 기간(Interval)로 쓸 수 없는 경우 (isMoment() 가 true 면서, isClosed() 도 true인 경우)
   */
  def isDegenerate: Boolean = isMoment && isClosed

  /**
   * 사용가능한 Interval 인가?
   */
  def isIntervalEnabled: Boolean = _intervalEnabled

  /**
   * 사용가능한 Interval 인지를 설정합니다.
   */
  def setIntervalEnabled(enabled: Boolean = true) { assertMutable; _intervalEnabled = enabled }

  override def hasStart =
    super.hasStart || !isStartClosed

  /**
   * 시작 시각
   */
  def getStartInterval: DateTime = _start

  /**
   * 시작 시각을 설정합니다.
   */
  def setStartInterval(newStart: DateTime) {
    assertMutable;
    assert(newStart.compareTo(_end) <= 0, "새로운 start가 end 보다 작아야 합니다.")
    _start = newStart
  }

  override def getStart: DateTime = if (_intervalEnabled && isStartOpen) _start.plus(1) else _start

  /**
   * 시작 시각의 개/폐구간 종류
   */
  def getStartEdge: IntervalEdge = _startEdge

  /**
   * 시작 시각의 걔/페구간 종류를 설정합니다.
   */
  def setStartEdge(edge: IntervalEdge) { assertMutable; _startEdge = edge }

  override def hasEnd = _end != TimeSpec.MaxPeriodTime || !isEndClosed

  /**
   * 완료 시각
   */
  def getEndInterval: DateTime = _end

  /**
   * 완료 시각을 설정합니다.
   */
  def setEndInterval(newEnd: DateTime) {
    assertMutable;
    assert(_start.compareTo(newEnd) <= 0, "end 는 start보다 크거나 같아야 합니다.")
    _end = newEnd
  }

  override def getEnd = if (_intervalEnabled && isEndOpen) _end.minus(1) else _end

  /**
   * 완료 시각의 개/폐구간 종류
   */
  def getEndEdge: IntervalEdge = _endEdge

  /**
   * 완료 시각의 개/폐구간 종류를 설정합니다.
   */
  def setEndEdge(edge: IntervalEdge) { assertMutable; _endEdge = edge }

  override def getDuration: Long = _end.getMillis - _start.getMillis

  override def setup(newStart: DateTime, newEnd: DateTime) {
    super.setup(newStart, newEnd)
    val (start, end) = Times.adjustPeriod(newStart, newEnd)
    _start = if (start != null) start else TimeSpec.MinPeriodTime
    _end = if (end != null) end else TimeSpec.MaxPeriodTime
  }

  def expandStartTo(moment: DateTime) {
    assertMutable
    if (_start.compareTo(moment) > 0) _start = moment
  }

  def expandEndTo(moment: DateTime) {
    assertMutable
    if (_end.compareTo(moment) < 0) _end = moment
  }

  def expandTo(moment: DateTime) {
    assertMutable
    expandStartTo(moment)
    expandEndTo(moment)
  }

  def expandTo(period: ITimePeriod) {
    assertMutable
    expandStartTo(period.getStart)
    expandEndTo(period.getEnd)
  }


  def shrinkStartTo(moment: DateTime) {
    assertMutable
    if (_start.compareTo(moment) < 0) _start = moment
  }

  def shrinkEndTo(moment: DateTime) {
    assertMutable
    if (_end.compareTo(moment) > 0) _end = moment
  }

  def shrinkTo(moment: DateTime) {
    assertMutable
    shrinkStartTo(moment)
    shrinkEndTo(moment)
  }

  def shrinkTo(period: ITimePeriod) {
    assertMutable
    shrinkStartTo(period.getStart)
    shrinkEndTo(period.getEnd)
  }

  def copy(): ITimeInterval = copy(TimeSpec.ZeroMillis)

  override def copy(offset: Long): ITimeInterval =
    TimeInterval(getStartInterval.plus(offset),
                 getEndInterval.plus(offset),
                 getStartEdge,
                 getEndEdge,
                 isIntervalEnabled,
                 isReadonly)


  override def reset() {
    super.reset()

    _intervalEnabled = true
    _startEdge = IntervalEdge.Closed
    _endEdge = IntervalEdge.Closed
  }

  override def getIntersection(other: ITimePeriod): ITimeInterval = {
    Guard.shouldNotBeNull(other, "other")
    val range = super.getIntersection(other)
    if (range == null) null
    else TimeInterval(range.getStart,
                      range.getEnd,
                      intervalEnabled = _intervalEnabled,
                      readonly = isReadonly)
  }

  override def getUnion(other: ITimePeriod): ITimeInterval = {
    Guard.shouldNotBeNull(other, "other")
    Times.getUnionRange(this, other)
  }

  override def format(formatter: ITimeFormatter): String =
    formatter.getInterval(_start, _end, _startEdge, _endEdge, getDuration)

  override def hashCode: Int =
    ScalaHash.compute(super.hashCode, _startEdge, _endEdge, _intervalEnabled)
}
